File: 3.x.x/crates-utopia/emails/src/validator/email_corporate.rs
use serde_json::Value;
use utopia_validators::{Validator, ValueType};

use crate::Email;

/// Validate that an email is corporate - not free or disposable (PHP `EmailCorporate`).
#[derive(Debug, Clone, Default)]
pub struct EmailCorporate;

impl EmailCorporate {
    /// Create a corporate-email validator.
    pub fn new() -> Self {
        Self
    }
}

impl Validator for EmailCorporate {
    fn description(&self) -> String {
        "Value must be a valid email address from a corporate domain".into()
    }

    fn is_array(&self) -> bool {
        false
    }

    fn value_type(&self) -> ValueType {
        ValueType::String
    }

    fn is_valid(&self, value: &Value) -> bool {
        let Some(raw) = value.as_str() else {
            return false;
        };
        Email::new(raw)
            .map(|email| email.is_valid() && email.is_corporate())
            .unwrap_or(false)
    }
}

File: packages/emails/src/Emails/Validator/EmailCorporate.php
<?php

namespace Utopia\Emails\Validator;

use Utopia\Emails\Email;
use Utopia\Validator;

/**
 * EmailCorporate
 *
 * Validate that an email address is from a corporate domain (not free or disposable)
 */
class EmailCorporate extends Validator
{
    /**
     * Get Description
     *
     * Returns validator description
     */
    public function getDescription(): string
    {
        return 'Value must be a valid email address from a corporate domain';
    }

    /**
     * Is valid
     *
     * Validation will pass when $value is a valid email address from a corporate domain
     *
     * @param  mixed  $value
     */
    public function isValid($value): bool
    {
        if (! is_string($value)) {
            return false;
        }

        try {
            $email = new Email($value);

            return $email->isValid() && $email->isCorporate();
        } catch (\Exception $e) {
            return false;
        }
    }

    /**
     * Is array
     *
     * Function will return true if object is array.
     */
    public function isArray(): bool
    {
        return false;
    }

    /**
     * Get Type
     *
     * Returns validator type.
     */
    public function getType(): string
    {
        return self::TYPE_STRING;
    }
}
